sleepecg/classification.py

```diff
@@ -0,0 +1,756 @@
+        `'wake-rem-nrem'`, `'wake-rem-light-n3'`, `'wake-rem-n1-n2-n3'`.
+    feature_extraction_params : dict[str, typing.Any]
+        The parameters passed to :func:`extract_features`, as a dictionary
+        mapping string parameter names to values. Should not include
+        `records` and `n_jobs`.
+    mask_value : int, optional
+        Only required for keras models, as passed to `prepare_data_keras`
+        and `keras.layers.Masking`, by default `None`.
+    classifiers_dir : str | pathlib.Path, optional
+        Directory in which the `.zip` file is stored. If `None`
+        (default), the value is be taken from the configuration.
+
+    See Also
+    --------
+    load_classifier : Load classifiers saved with this function
+    """
+    if classifiers_dir is None:
+        classifiers_dir = get_config('classifiers_dir')
+
+    target_file = Path(classifiers_dir).expanduser() / name
+
+    model_type = model.__module__.split('.')[0]
+    classifier_info = {
+        'model_type': model_type,
+        'stages_mode': stages_mode,
+        'feature_extraction_params': feature_extraction_params,
+    }
+    if mask_value is not None:
+        classifier_info['mask_value'] = mask_value
+
+    with TemporaryDirectory() as tmpdir:
+        with open(f'{tmpdir}/info.yml', 'w') as infofile:
+            yaml.dump(classifier_info, infofile)
+
+        if model_type == 'keras':
+            model.save(f'{tmpdir}/classifier')
+        else:
+            raise ValueError(f'Saving model of type {type(model)} is not supported')
+
+        shutil.make_archive(target_file, 'zip', tmpdir)
+
+
+@dataclass
+class SleepClassifier:
+    """
+    Store a sleep classifier model and metadata.
+
+    Attributes
+    ----------
+    model : typing.Any
+        The classification model, should have `fit` and `predict` methods.
+    stages_mode : str
+        Identifier of the grouping mode. Can be any of `'wake-sleep'`,
+        `'wake-rem-nrem'`, `'wake-rem-light-n3'`, `'wake-rem-n1-n2-n3'`.
+    feature_extraction_params : dict[str, typing.Any]
+        The parameters passed to :func:`extract_features`, as a dictionary
+        mapping string parameter names to values. Does not include
+        `records` and `n_jobs`.
+    model_type : str
+        A string identifying the model type, e.g. `'keras'` or `'sklearn'`.
+        This is used by :func:`stage` to determine how to perform sleep
+        stage predictions.
+    mask_value : int, optional
+        Only required for models of type `'keras'`, as passed to
+        `prepare_data_keras` and `keras.layers.Masking`, by default `None`.
+    source_file : str, optional
+        The file from which the classifier was loaded using
+        :func:`load_classifier`, by default `None`.
+    """
+
+    model: Any
+    stages_mode: str
+    feature_extraction_params: Dict[str, Any]
+    model_type: str
+    mask_value: Optional[int] = None
+    source_file: Optional[str] = None
+
+    def __str__(self) -> str:
```

Comment:
Do we also want/need a `__repr__`?

Reply:
I'd prefer not to, as it would be difficult to follow what the [Python docs](https://docs.python.org/3/library/functions.html#repr) here.

Reply:
Yes, but it could at least be more informative than the default memory address repr. For interactive use it's quite handy to just type the name of the object instead of an explicit print.

Reply:
But would a more conventional one-line `__repr__` also work, e.g.

```
f'SleepClassifier(model, {self.stages_mode.upper()}, feature_extraction_params, model_type={self.model_type}'
```

etc.?

What does the default `repr` look like? Do you have an example?

Reply:
that's what the default repr looks like:
```
SleepClassifier(model=<keras.engine.sequential.Sequential object at 0x000001DF1CF33108>, stages_mode='wake-sleep', feature_extraction_params={'feature_selection': ['hrv-time', 'hrv-frequency', 'recording_start_time', 'age', 'gender'], 'lookback': 240, 'lookforward': 270, 'max_nans': 0.5, 'max_rri': 2, 'min_rri': 0.3}, model_type='keras', mask_value=-1, source_file='c:\\users\\florian\\desktop\\masterarbeit\\code\\sleepecg\\sleepecg\\classifiers\\ws-gru-mesa.zip')
```

I didn't realize it contains that much information (thanks to `SleepClassifier` being a dataclass), so I'd actually vote for keeping the `__str__` as is, and leaving the default `__repr__`.

Reply:
Yes, that's probably the best solution for now. I thought about making the repr a bit more concise, but there's no way it could contain all of this information. In MNE, we'd ignore the part that the repr should be a string to create the object. Instead, we only show the most important bits of information, so this could be:

```
<SleepClassifier | keras, wake-sleep, ws-gru-mesa.zip>
```

And `__str__` could then show more (all?) information. WDYT?

Reply:
> ```
> <SleepClassifier | keras, wake-sleep, ws-gru-mesa.zip>
> ```

I like that! So to summarize, we keep the current implementation of `__str__`, and add a `__repr__` in the style above?

Reply:
I added the `__repr__` in 8e93dd6, with a two modifications:
- `SleepClassifier.source_file` is now a `pathlib.Path`, to make getting the name cleaner
- `stages_mode` comes before `model_type`, as I think it's more important